### DISCUSSION TITLE
Network handling

### DISCUSSION THREAD

**Discussion by datablitz7** [htpcBeginner/docker-traefik#119] [Q&A]
Is there a particular reason for creating the proxy network outside the docker-compose file? It seems counter-intuitive to have an additional step. Personally I use the default bridge that is implicitly created up docker-compose up, but you can still name it and stand it up all in the yaml file:

```
networks:
  t2_proxy:
    name: t2_proxy
    driver: bridge
  docker_socket_proxy:
    name: docker_socket_proxy
    driver: bridge
```

**Comment by gzecchi** [comment]
Hi @datablitz7,

the reason being of creating the proxy network outside the docker-compose file is to select a custom subnet, that doesn't overlap with your existing networks, instead of letting docker choosing it.

In this way, you are able to specify the IP address of any container and reference it in some specific configuration that doesn't allow DNS reference. 

If you select the Bridge mode, all the containers can talk to each other and can reach (and be reached by) any outside resource without any limitation.
For security reasons, you probably don't want that anyone in your network can run containers so you use the docker-socket-proxy that is connected to an "internal" connection, in this way only the containers that run in the same host and are connected to that network can access to the docker daemon.

GZ

**Reply by datablitz7** [reply]
Hi @gzecchi,

I am not debating the isolation of networking and not suggesting using the default docker bridge, but merely to leverage the fact that a docker-compose stack get its own network bridge implicitly. If you do not define one, you will get a network named "default" with a  random subnet. You can still define it though within the docker-compose, selecting both a different name and the subnet of choice:

```
networks:
  t2_proxy:
    name: t2_proxy
    driver: bridge
    ipam:
      config:
        - subnet: 192.168.50.0/24
```

What I am therefore asking, is whether there is in fact a reason to rely on separately creating a docker bridge outside the docker compose file, as it just seems to make the entire setup a little less portable.

**Reply by gzecchi** [reply]
Hi @datablitz7,
thank you for the explanation.

I think we could integrate it into the docker-composer file.

GZ

**Reply by SimpleHomelab** [reply]
Interesting. I am going to try and implement this. 

EDIT: This works and is much cleaner. I will roll this out in future.

**Reply by tehniemer** [reply]
just a note, `name:` only works with compose v3.5 and higher.

**Reply by datablitz7** [reply]
Indeed, but since the docker-compose files in this repo are for 3.7, I thought it could go without saying.

**Reply by datablitz7** [reply]
> Interesting. I am going to try and implement this.
> 
> EDIT: This works and is much cleaner. I will roll this out in future.

I have put this together in a PR. https://github.com/htpcBeginner/docker-traefik/pull/126

**Reply by android-808** [reply]
At the moment I tend to test containers in a separate docker-compose. This way I can leave the normal file and services alone and running and just test the new services on their own. If the network is defined in the compose file, will this still work with this config or is the network then only available to services in the main compose file?

**Reply by PrplHaz4** [reply]
> 
> 
> At the moment I tend to test containers in a separate docker-compose. This way I can leave the normal file and services alone and running and just test the new services on their own. If the network is defined in the compose file, will this still work with this config or is the network then only available to services in the main compose file?

You can use external networks in the other test compose to reference your "prod" compose network:
https://docs.docker.com/compose/networking/#use-a-pre-existing-network

```yml
networks:
  prodnet:
    external:
      name: t2_proxy
```

**Comment by SimpleHomelab** [comment]
I implemented this in the latest commit: ec9b3b2a8d3176f304de1f4340160256eca8cfd2
